### PR TITLE
Add GetOpt module to decode the command line options supplied to Flix's main. 

### PR DESCRIPTION
**Pull request by stephentetley**
I've ported Haskell's `System.Console.GetOpt` module to Flix so we can decode arguments supplied to main.

Included in this PR is a test file and an example of writing a command arguments processor.

The test file is a bit unusual as the GetOpt module has a very small API - one major function (`getOpt`) plus a lower level version (`getOpt1`) that is same function but has a different return type. Alongside these two functions there are three public datatypes. Because we have one main function, the tests are grouped to exercise its various functionality and there is a fair amount of support code at the top of the test file.

I've added the function `preprocess` (which Haskell's GetOpt doesn't have) so we can preprocess the original args before decoding them - this is principally so we can handle file names with spaces, rejoining them if they are enclosed in quotes  (I'm mostly a Windows user so this is quickly a problem). There are extra tests in the test file for `preprocess` function.

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Thanks Stephen! Looks awesome 👍  

Give me some time to review.

I am very happy that you both added a lot of tests and an example for the website :)

**Comment by stephentetley** [comment]
I've push a small improvement to `preprocess` so it will recognize if a single token opens and ends a quote. In this case the quote marks would be superfluous in the user's input, but I think we should be able to recognize them anyway.

File: main/src/library/GetOpt.flix [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+        }
+
+
+    ///
+    /// Return a formatted string describing the usage of the command.
+    ///
+    /// Typically this output is used by "--help"
+    /// 
+    pub def usageInfo(header: String, optionDescriptors: List[OptionDescr[a]]): String = 
+        let paste = (x, y, z) -> "  ${x}  ${y}  ${z}";
```
Comment (magnus-madsen) [inline main/src/library/GetOpt.flix:94]: Are there two spaces here? is that intended?

**Comment by magnus-madsen** [comment]
Hello Stephen,

I am happy to merge this in as-is. I do have some high-level comments/questions:

- **Purity**: The implementation looks purely functional to me, yet many functions are marked impure. I assume this is because reading from an array is impure (and so pollutes everything with impurity). My questions is whether it would be possible (or even desirable) to either (a) immediately convert `args` to a list and work with that or (b) cast some of the functions to pure (since they are effectively pure).

- *Naming*: I worry that some names might conceptually clash with other naming in the standard library. What I mean is that Flix has `Option` (whereas Haskell has `Maybe`) and so the use of things like `OptionKind` and `OptionDescr` might be confusing. Moreover, e.g. in #1679, we might have to resort to naming functions with `option` as part of their name. For example, `reduceLeftOption` adding to the confusion. My question is whether (a) there is a word we could substitute for Option in this context? Perhaps Flag or Argument or Parameter? My second question is (b) if the names follow the names of the Haskell library- because if it does I can see the value in preserving the naming scheme used there. I am not beyond pragmatism :)

- *Packages*: At some point (hopefully sooner rather than later) Flix will have a package system. When that happens it will make sense to split up the standard library a bit. Command line argument parsing is *potentially* one of the things it could make sense to factor out (an argument could also be made to keep it in). My point is that while I am happy to accept it in the std. lib. today, it might be that in one year we will want to "move it out" again. Or at least have the discussion at that point.

Let me know your thoughts and then we can merge it in soon.

**Comment by magnus-madsen** [comment]
!!!benchmark

File: main/src/library/GetOpt.flix (hunk 1 above)
Comment (stephentetley) [reply main/src/library/GetOpt.flix:94]: There are two spaces in the Haskell original, so I kept the same formatting.

**Comment by stephentetley** [comment]
Hi Magnus, thanks for the comments...

+ **Purity** I'm happy to change if you prefer, but not troubled by the the impurity myself. Essentially I see the functions being called from `main` which is impure, as it's so close to `main` and performs something very specific I'd want to use the type that main uses. If we decide `preprocess` needs improvement as it is very simplistic there's a good case to shift to lists.

+ **Naming** I've confused `OptionDescr` for `Option` in code myself when I was writing another style of decoder. There does seem to be a general convention via `getopt` in Unix-land that _commands_ take a list of _options_ (and _non-options_ e.g filenames) and _options_ themselves may have _arguments_. I changed `OptionDescr` from `OptDescr` in Haskell because Flix doesn't have a convention of using abbreviations (the full name `OptionDescriptor` did seem too long), but I didn't change `ArgDescr` as "arg" / "args" seems a much more common abbreviation. I'm happy to change `OptionDescr` back to to `OptDescr`. If we try to avoid "Option" altogether I think we might have an API that all newcomers that have used a GetOpt-like library will find confusing.

+ **Packages** As Flix grows I can envisage other command option processors, as has happened in the Haskell world. It wouldn't be a bad thing if GetOpt is moved out of the standard library as long as it is still "findable" which may imply something like Hackage - a package repository - as well as a package system.

**Comment by magnus-madsen** [comment]
Merged! Thanks Stephen!

**Pull request merged**